src/contrasts.jl

```diff
@@ -466,49 +474,79 @@ expressed as a hypothesis matrix, but it is not obvious just from looking at the
-    hypotheses::Matrix
-    contrasts::Matrix
-    base::Nothing
-    levels::Union{Vector,Nothing}
-
-    function HypothesisCoding(hypotheses, base, levels)
+mutable struct HypothesisCoding{T<:AbstractMatrix, S<:AbstractMatrix} <: AbstractContrasts
+    hypotheses::T
+    contrasts::S
+    levels::Union{AbstractVector,Nothing}
+    labels::AbstractVector
+
+    function HypothesisCoding(hypotheses::T, levels, labels) where {T}
+        labels == nothing &&
+            throw(ArgumentError("must specify contrast labels with " *
+                                "HypothesisCoding(; labels=) or " *
+                                "HypothesisCoding(Dict(label1=>hyp1, label2=>hyp2, ...))"))
         contrasts = pinv(hypotheses)
+        S = typeof(contrasts)
         check_contrasts_size(contrasts, levels)
-        new(hypotheses, contrasts, base, levels)
+        new{T,S}(hypotheses, contrasts, levels, labels)
     end
 end
 
-HypothesisCoding(mat::Matrix; levels=nothing) =
-    HypothesisCoding(mat, nothing, levels)
+HypothesisCoding(mat::AbstractMatrix; levels=nothing, labels=nothing) =
+    HypothesisCoding(mat, levels, labels)
+
+"""
+    HypothesisCoding(hypotheses::Dict[; labels=, levels=])
+
+Specify hypotheses as `label=>hypothesis_vector` pairs.  If labels are specified
+via keyword argument, the hypothesis vectors will be concatenated in that order.
+"""
+function HypothesisCoding(
+    hypotheses::Dict{<:Any,<:AbstractVector};
+    labels=collect(keys(hypotheses)),
+    levels=nothing
+)
+    !isempty(symdiff(keys(hypotheses), labels)) &&
+        throw(ArgumentError("labels mismatch between hypotheses " *
+                            "($(collect(keys(hypotheses)))) " *
+                            "and labels kw arg ($labels)"))
+    
+    mat = reduce(hcat, hypotheses[label] for label in labels)'
+    HypothesisCoding(mat; labels=labels, levels=levels)
+end
+
 
 function contrasts_matrix(C::HypothesisCoding, baseind, n)
     check_contrasts_size(C.contrasts, n)
     C.contrasts
 end
 
+termnames(C::HypothesisCoding, levels::AbstractVector, baseind::Int) =
+    something(C.labels, levels[1:length(levels) .!= baseind])
 
 """
-    StatsModels.ContrastsCoding(mat::Matrix[, base[, levels]])
+    StatsModels.ContrastsCoding(mat::AbstractMatrix[, base[, levels]])
 
 Coding by manual specification of contrasts matrix. For k levels, the contrasts
 must be a k by k-1 Matrix.  The contrasts in this matrix will be copied directly
 into the model matrix; if you want to specify your contrasts as hypotheses (i.e., 
 weights assigned to each group's cell mean), you should use 
 [`HypothesisCoding`](@ref) instead.
 """
-mutable struct ContrastsCoding <: AbstractContrasts
-    mat::Matrix
-    base::Any
+mutable struct ContrastsCoding{T<:AbstractMatrix} <: AbstractContrasts
+    mat::T
     levels::Union{Vector,Nothing}
 
-    function ContrastsCoding(mat, base, levels)
+    function ContrastsCoding(mat::T, levels) where {T}
```

Comment:
```suggestion
    function ContrastsCoding(mat::T, levels) where {T<:AbstractMatrix}
```
Also same remark about using a keyword argument here.

Reply:
...and same thing about there being a separate kwargs constructor defined outside.